### PR TITLE
More gas oracles

### PR DESCRIPTION
**Pull request by recmo**
## Motivation

This adds the `ProviderOracle` that allows runtime selection between provider gas methods and gas oracles. Such runtime selection is very useful for building stacks that work on multiple chains, such as production and local test.

~~It also adds a `default_for_chain` constructor that looks at the `chain_id` and tries to build a sensible default. This is a bit proof-of-concept. It would be nice to add a builder pattern where you can provide api-keys.~~

Architecturally this all feels a bit off to me. `ProviderOracle` exists basically because `GasOracle` and `Middleware` overlap in functionality. The `Provider` implementation also does quite a bit of magic under the hood to get EIP1559 estimates. IMHO it would be cleaner if this was all done through something like GasOracle's. But simply factoring it out doesn't work because gas prices are required to sign transactions, and at least currently Provider provides a complete implementation. Makes me think the Middleware really should be broken up in separate parts relating to the transaction lifecycle (gas pricing, gas estimation, gas escalation, nonce, signing, broadcasting, monitoring, etc). (Also seems related to #592)

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Updated the changelog

### CODE REVIEW

**Comment by mattsse** [comment]
cc @oliver-giersch

File: ethers-middleware/src/gas_oracle/provider_oracle.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+    async fn fetch(&self) -> Result<U256, GasOracleError> {
+        self.provider
+            .get_gas_price()
+            .await
+            .map_err(|err| GasOracleError::ProviderError(Box::new(err)))
+    }
+
+    async fn estimate_eip1559_fees(&self) -> Result<(U256, U256), GasOracleError> {
+        // TODO: Allow configuring different estimation functions.
```
Comment (mattsse) [inline ethers-middleware/src/gas_oracle/provider_oracle.rs:34]: perhaps something like a closure that returns a future?

**Review by mattsse — APPROVED** [review]
lgtm

thanks for this, definitely something we need to keep in mind for #592

**Comment by gakonst** [comment]
> Makes me think the Middleware really should be broken up in separate parts relating to the transaction lifecycle (gas pricing, gas estimation, gas escalation, nonce, signing, broadcasting, monitoring, etc). (Also seems related to https://github.com/gakonst/ethers-rs/issues/592)

This is not a bad idea, although it does prevent you from join'ing all the futures for gas price + gas estimation + nonce, which is a nice performance boost. Do you wanna drop any thoughts on the architecture in the issue?

**Comment by gakonst** [comment]
(Think this needs some fixes around WASM)

**Comment by recmo** [comment]
I removed the `default_for_provider` function. It was constructing a reqwest client with timeout, which is not available on WASM, causing the build to fail. I split the `LogQueryError` out to it's own PR #1615 .

**Review by gakonst — APPROVED** [review]
sweet

**Pull request merged**